flink-table/flink-table-code-splitter/src/main/java/org/apache/flink/table/codesplit/BlockStatementGrouper.java

```diff
@@ -0,0 +1,474 @@
+                            .collect(Collectors.toList());
+
+            groupStrings.put(group.getKey(), collectedStringGroups);
+        }
+
+        return new RewriteGroupedCode(rewriter.getText(), groupStrings);
+    }
+
+    private static class BlockStatementGrouperVisitor {
+
+        private final Map<String, Pair<TokenStreamRewriter, List<LocalGroupElement>>> groups =
+                new HashMap<>();
+
+        private final long maxMethodLength;
+
+        private final String parameters;
+
+        private int counter = 0;
+
+        private BlockStatementGrouperVisitor(long maxMethodLength, String parameters) {
+            this.maxMethodLength = maxMethodLength;
+            this.parameters = parameters;
+        }
+
+        public void visitStatement(
+                StatementContext ctx, String context, TokenStreamRewriter rewriter) {
+
+            if (ctx.getChildCount() == 0) {
+                return;
+            }
+
+            // For these statements here we want to process all "branches" separately, for example
+            // TRUE and FALSE branch of IF/ELSE block.
+            // each statement can be rewritten and extracted.
+            if (ctx.WHILE() != null || ctx.IF() != null || ctx.ELSE() != null) {
+                for (StatementContext statement : ctx.statement()) {
+                    if (shouldExtract(statement)) {
+                        String localContext = String.format("%s_%d", context, counter++);
+                        groupBlock(statement, localContext, rewriter);
+                    }
+                }
+            } else {
+                // The block did not start from IF/ELSE/WHILE statement
+                if (shouldExtract(ctx)) {
+                    groupBlock(ctx, context, rewriter);
+                }
+            }
+        }
+
+        // Group continuous block of statements together. If Statement is an IF/ELSE/WHILE,
+        // its body can be further grouped by recursive call to visitStatement method.
+        private void groupBlock(
+                StatementContext ctx, String context, TokenStreamRewriter rewriter) {
+            int localGroupCodeLength = 0;
+            List<LocalGroupElement> localGroup = new ArrayList<>();
+            for (BlockStatementContext bsc : ctx.block().blockStatement()) {
+
+                StatementContext statement = bsc.statement();
+                if (statement.IF() != null
+                        || statement.ELSE() != null
+                        || statement.WHILE() != null) {
+                    String localContext = context + "_rewriteGroup" + this.counter++;
+
+                    CommonTokenStream tokenStream =
+                            new CommonTokenStream(
+                                    new JavaLexer(
+                                            CharStreams.fromString(
+                                                    CodeSplitUtil.getContextString(statement))));
```

Comment:
You say "In my case we parse class once" in a previous comment, but here you repeatedly and recursively parse the blocks.

Reply:
Please remember that that comment was created for previous implementation, where I was using multiple instances of Splitter and Grouper visitors. I refactor both classes so now they are using recursive calls to `visit...` methods. 

Regarding the "repeatedly parse" part.
Statement is processed by Splitter and Grouper once. I'm not sure how it looks from the Antlr4 but I guess it will be lie so:
1. Initial parse for Splitter (part of original block statement to rewrite)
2. Initial parser for Grouper  (part of original block statement to rewrite)
3. possibly a third time when we create a new rewriter instane for `RewriteContextGroupElement`

I think that my main point was that in my implementation I don't parse entire class again every time when `IfStatementRewriter` is doing another round to process extracted methods. At least this is what I think is happening in `IfStatementRewriter` and that was the reason you had

```
        while (visitor.hasRewrite()) {
            visitor = new IfStatementVisitor(rewriterCode);
            rewriterCode = visitor.rewriteAndGetCode();
        }
```

I think that in my case, the number of times that statement will be parsed by antlr or processed by Splitter/Grouper is constant, whereas in `IfStatementRewriter` it depends on the number of extracted methods.

But I want to be crystal clear here. I dont want to provoke a discussion about the performance especially when you said that with original approach you never had issue regarding the performance - I get that. When I start implementing my solution this was not my goal. Its a side effect of the design approach I took  i think.